text/Issue 8 - Plug-in based architecture.md

```diff
@@ -0,0 +1,265 @@
+come at some cost performance cost (typically, they may need to reparse the
+same value several times from JSON or Binary).
+
+
+### Implementation
+
+```rust
+trait Message: Sized {
+  /// Developer-oriented human-readable name, used mainly for error messages.
+  ///
+  /// For instance, `bool`, `positive number`, `temperature`.
+  fn name() -> String;
+
+  /// Values carried by messages of this type.
+  type Type: Any;
+
+  /// A mechanism used to serialize values of this type.
+  type Serializer: Serializer<Source = Self::Type>;
+
+  /// A mechanism used to deserialize values of this type.
+  type Deserializer: Deserializer<Target = Self::Type>;
+}
+
+/// A mechanism used to serialize values of a given type.
+trait Serializer {
+  type Source;
+
+  /// Serialize a value of type `Source` to JSON. If the value has binary
+  /// components, they should be placed in `binary`.
+  fn serialize(source: &Self::Source, binary: &mut BinaryComponents) -> Result<JSON, SerializeError>;
+}
+trait Deserializer {
+  type Target;
+
+  /// Deserialize a value of type `Target`. If the value has binary components,
+  /// they can be fetched from `binary`.
+  fn deserialize(source: &JSON, binary: &BinaryComponents) -> Result<Self::Target, ParseError>;
+}
+```
+
+Note that `BinaryComponents` are taken from ongoing work on https://github.com/fxbox/taxonomy/issues/37
+and are not specific to this RFC.
+
+
+## Channel features
+
+We replace `ChannelKind` with `Feature`. As `ChannelKind`, the role of a
+`Feature` is to allow an application to discover channels (respectively service)
+that implement a given feature (respectively feature set), and to standardize
+the (de)serialization format for this channel.
+
+A crate `standardized` provides a library of standard `Feature`s to cover a number
+of uses:
+
+- garage door opened;
+- lock unlocked;
+- oven temperature;
+- ...
+
+Use of a `Feature` is decided per-channel, by the Adapter developer.
+
+Third-party developers are encouraged to pick a standardized `Feature`, but
+are not forced to. Additional instances of `Feature` can easily be published
+by third-party developers to third-party crates.
+
+### Implementation
+
+
+```rust
+struct Feature<T> where T: Message {
+  /// Keys used to discover this feature.
+  ///
+  /// For instance: `["x-oven-temperature", "oven-temperature"]`, where
+  /// `x-oven-temperature` was the name used prior to standardization and
+  /// `oven-temperature` is the standardized name.
+  pub keys: Vec<String>,
+
+  phantom: PhantomData<T>
+}
+```
+
+
+## Propagating the changes
+
+We modify the definition of channels to use `Feature<T>` instead of
+`ChannelKind`. Otherwise, this definition is unchanged. We similarly modify
+`add_getter`/`add_setter` to use this new definition of `Channel<T>`.
+
+In the implementation of `Adapter`, type `Value` is replaced by `Any`.
+
+In the Taxonomy API, type `Value` is replaced by `JSON`, with a companion `BinaryParts`.
```

Comment:
No, that's not the reason (although that would be an interesting application, too).

The reason for `BinaryParts` is that binary embedded in JSON is a pretty bad idea. So, when the foxbox exposes it through the REST API, we need to publish it either as a multipart, as a short-lived url, or by some other mean yet undecided. Similarly, when the foxbox receives it through the REST API, it pretty much needs to be either multipart or form-encoded.

In either case, this means that we need to collect both JSON and Binary.

> Is there a spec for the JSON to refer to individual binary parts (as in { "item": "hey-fetch-the-contents-of-binary-1-here" }) or are they just known attachments?

I'm not aware of any spec.

> I wouldn't be opposed to having a class like Value that encapsulates both the JSON and BinaryParts.

Having implemented a prototype of the serialization-to-json-and-multipart, I feel that it's simpler to produce a JSON and fill a binary store, rather than attempting to build both and somehow keep the "hey-fetch-the-contents-of-binary-1-here" synchronized.